### DISCUSSION TITLE
Listing templates - Make provider is always pending when the `overseer.template.list()` function is called

### DISCUSSION THREAD

**Discussion by georgev93** [stevearc/overseer.nvim#384] [Q&A]
Because I'd really like to be able to quickly run a "test" task and a "build" task for a given project, I implemented a sort of "run default test" and "run default build" keybindings via a small plugin (wrapper-like) with overseer ([buried in my dots](https://github.com/georgev93/dotfiles/tree/main/nvim/lua/local_plugins/test_picker)). This works great! I'm even able to specify default templates for building or testing by name in the opts when I load the plugin. Great times.

That is to say, it works most of the time. If there is a makefile present, it's not able to enumerate a list of templates if there is a Makefile. My "plugin" works by using a portion of overseer.nvim not exposed by the API (dangerous, I know. Also, as an aside, I think this would be a great thing to expose!): overseer.template.list(). When I call it and there is a Makefile present, it will fail and say: `[ERROR] Listing templates timed out. Pending providers: { "make" }`. I can wait for any amount of time, always the same result. But the second time I call the list() function, everything works as intended.

And for the record, this is the observed behavior even if the Makefile is all commented out. And I do have make installed on my machine.

If I change the `template_cache_threshold` to 0, it fails every time. So this would seem to indicate that the list() function is able to put the right templates in the cache, but not in time to return from the function.

I'm trying to wrap my head around what's happening in list(), but it's slow going. Does anyone know the inter-workings of this function a bit more intimately and would be willing to help me get to the bottom of this any faster?

Also, suggestions for alternatives to this list() function are welcomed! And if anyone likes this enough to pull this into overseer.nvim, I'd be happy to work with you to make that happen! Not sure if this is a feature anyone else would want, though.

**Comment by stevearc** [comment]
I have no idea why it would consistently timeout, especially if it works properly when called from `:OverseerRun`. For your use case, I would generally recommend to use `overseer.run_template`. Is there a reason that doesn't work for you?
